query-algebrizer/src/clauses/predicate.rs

```diff
@@ -81,13 +94,181 @@ impl ConjoiningClauses {
+        //     },
+        //     FnArg::Constant(NonIntegerConstant::Text(s)) => {
+        //         QueryValue::TypedValue(TypedValue::typed_string(s.as_str()))
+        //     },
+        //     _ => bail!(ErrorKind::InvalidArgument(where_fn.operator.clone(), "string".into(), 2)),
+        // };
+
+        // // TODO: should we build the FQA in ::Matches, preventing nonsense like matching on ::Rowid?
+        // let constraint = ColumnConstraint::Matches(QualifiedAlias(fulltext_values_alias.clone(), Column::Fulltext(FulltextColumn::Text)), search);
+        // self.wheres.add_intersection(constraint);
+
+        let mut values: Vec<Vec<TypedValue>> = Default::default();
+
+        let mut projected: Vec<Variable> = Default::default();
+
+        // TODO: binding-specific error messages.
+        match where_fn.binding {
+            Binding::BindScalar(var) => {
+                // if bindings.len() > 4 {
+                //     bail!(ErrorKind::InvalidNumberOfArguments(where_fn.operator.clone(), bindings.len(), 4));
+                // }
+                // bindings.into_iter()
+
+                if self.bound_value(&var).is_some() || self.input_variables.contains(&var) {
+                    // That's a paddlin'!
+                    bail!(ErrorKind::InvalidArgument(where_fn.operator.clone(), "illegal bound variable".into(), 999))
+                }
+
+                let constant: TypedValue = match args.next().unwrap() {
+                    FnArg::EntidOrInteger(x) => {
+                        TypedValue::Long(x)
+                    },
+                    FnArg::Ident(x) => {
+                        TypedValue::Keyword(Rc::new(x))
+                    },
+                    FnArg::Constant(NonIntegerConstant::Boolean(x)) => TypedValue::Boolean(x),
+                    FnArg::Constant(NonIntegerConstant::BigInteger(x)) => unimplemented!(),
+                    FnArg::Constant(NonIntegerConstant::Float(x)) => TypedValue::Double(x),
+                    FnArg::Constant(NonIntegerConstant::Text(x)) => TypedValue::String(x),
+                    FnArg::Variable(_) | FnArg::SrcVar(_) => panic!("bad things happened"),
+                    FnArg::Collection(_) => panic!("more bad things happened"),
+                };
+
+                projected.push(var.clone());
+                self.constrain_var_to_type(var.clone(), constant.value_type());
+
+                values.push(vec![constant]); // 1x1.
+
+                // // TODO: this optimization.
+                // self.value_bindings.insert(var.clone(), constant.clone());
+
+                // let score_alias = QualifiedAlias(datoms_table_alias.clone(), Column::Fixed(DatomsColumn::Tx));
+            },
+            _ => bail!(ErrorKind::InvalidArgument(where_fn.operator.clone(), "bindings".into(), 999)),
+        };
+
+        let projection: BTreeSet<Variable> = projected.into_iter().collect();
```

Comment:
See comment I left on Emily's PR; we should probably make the various var lists in the parser all spit out `BTreeSet<Variable>` and err on dupes.